File: Data Structures & Algorithms/is-palindrome/submission-0.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int left = 0, right = s.length() - 1;

        while(left < right)
        {
            while(left < right && !isASCII(s[left]))
            {
                left++;
            }

            while(right > left && !isASCII(s[right]))
            {
                right--;
            }

            if(tolower(s[left]) != tolower(s[right])) return false;

            left++;
            right--;
        }

        return true;
    }

    bool isASCII(char c){
        return ((c >= 'A' && c <= 'Z') ||
                (c >= 'a' && c <= 'z') ||
                (c >= '0' && c <= '9'));
    }
};

File: Data Structures & Algorithms/path-sum/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        sum = 0
        def canReachNode(root, sum, target):
            if not root:
                return False
            sum += root.val

            if not root.left and not root.right and sum == target:
                return True
            if canReachNode(root.left, sum, target):
                return True
            if canReachNode(root.right, sum, target):
                return True
            
            sum -= root.val
            return False
        
        return canReachNode(root, sum, targetSum)